File: app/src/main/java/me/zhanghai/android/files/provider/linux/LocalLinuxWatchKey.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.linux;

import androidx.annotation.NonNull;
import me.zhanghai.android.files.provider.common.AbstractWatchKey;

class LocalLinuxWatchKey extends AbstractWatchKey {

    private final int mWd;

    private boolean mValid = true;

    LocalLinuxWatchKey(@NonNull LocalLinuxWatchService watchService, @NonNull LinuxPath path,
                       int wd) {
        super(watchService, path);

        mWd = wd;
    }

    int getWatchDescriptor() {
        return mWd;
    }

    @Override
    public boolean isValid() {
        synchronized (mLock) {
            return mValid;
        }
    }

    void setInvalid() {
        synchronized (mLock) {
            mValid = false;
        }
    }

    @Override
    public void cancel() {
        synchronized (mLock) {
            if (mValid) {
                getWatchService().cancel(this);
            }
        }
    }

    @NonNull
    @Override
    public LinuxPath watchable() {
        return (LinuxPath) super.watchable();
    }

    @NonNull
    @Override
    protected LocalLinuxWatchService getWatchService() {
        return (LocalLinuxWatchService) super.getWatchService();
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/linux/LocalLinuxWatchKey.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.linux

import me.zhanghai.android.files.provider.common.AbstractWatchKey

internal class LocalLinuxWatchKey(
    watchService: LocalLinuxWatchService,
    path: LinuxPath,
    val watchDescriptor: Int
) : AbstractWatchKey(watchService, path) {
    override val watchService: LocalLinuxWatchService
        get() = super.watchService as LocalLinuxWatchService

    private var isValid = true

    override fun isValid(): Boolean {
        synchronized(lock) { return isValid }
    }

    fun setInvalid() {
        synchronized(lock) { isValid = false }
    }

    override fun cancel() {
        synchronized(lock) {
            if (isValid) {
                watchService.cancel(this)
            }
        }
    }

    override fun watchable(): LinuxPath = super.watchable() as LinuxPath
}
